### PR TITLE
Rendering Optimisations - Part 2

### LINKED ISSUE #259: [Rendering] Rendering optimisations v2
## Background
After getting significant gains in #226, there's still much more juice we can squeeze out of the renderer.

The name of the game is _bandwidth_. We **must** ensure that we pull as little data down as possible in each draw.

## TODO
- [ ] #357
- [x] Use instancing (Qualcomm recommends this: https://developer.qualcomm.com/sites/default/files/docs/adreno-gpu/developer-guide/gpu/best_practices_vertex.html#consider-geometry-instancing)
- [ ] #356
- [ ] #274 
- [ ] https://www.unrealengine.com/en-US/blog/physically-based-shading-on-mobile
- [ ] https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_texture_transform/README.md

### PR DESCRIPTION
**Pull request by kanerogers**
## Background
Closes #259, #357, #356, #274

This is a *very* big PR. It touches almost all of Hotham's internals, and makes many breaking changes.

However, it also *dramatically* improves the performance of almost all Hotham applications. While I would have preferred to split this PR into multiple chunks, strong time pressure and the nature of inter-connected improvements made that incredibly difficult.

## TODO
- [x] Run through the full test suite and create issues for any problems that arise
- [x] Test all examples on-device

### CODE REVIEW

**Comment by kanerogers** [comment]
First bug is with FFR - there's no support for this in the simulator, obviously. I could potentially add some support for this, but it really doesn't make much sense. Instead I'll see if I can make it optional behind a `target_os` flag

**Comment by kanerogers** [comment]
I've *mostly* got this working:

- All tests pass
- Crab Saber, Simple Scene and friends all work fine on simulator

Next up is testing on-device and looking for performance regressions!

**Comment by kanerogers** [comment]
Android now builds! Just needs to be verified.

**Comment by kanerogers** [comment]
Okay! Everything works fine, no performance regressions on The Station. All is well! 🎉

**Comment by rasmusgo** [comment]
The gui render doesn't look good. The color of the ferris crab is broken, see `test_assets/render_draw_gui.jpg`
image

File: Cargo.toml [hunk 1]
```diff
@@ -1,13 +1,15 @@
 [workspace]
 members = [
-    "hotham",
-    "hotham-simulator",
-    "examples/shared",
-    "examples/simple-scene",
-    "examples/complex-scene",
-    "examples/custom-rendering",
```
Comment (rasmusgo) [inline Cargo.toml:8]: Custom rendering example was removed by mistake here.

Comment (rasmusgo) [reply Cargo.toml:8]: And doesn't compile if added...

File: hotham/src/contexts/physics_context.rs [hunk 2]
```diff
@@ -20,9 +19,9 @@ pub struct PhysicsContext {
     pub narrow_phase: NarrowPhase,
     pub rigid_bodies: RigidBodySet,
     pub island_manager: IslandManager,
-    pub collision_recv: Receiver<CollisionEvent>,
-    pub contact_force_recv: Receiver<ContactForceEvent>,
-    pub event_handler: ChannelEventCollector,
+    // pub collision_recv: Receiver<CollisionEvent>,
+    // pub contact_force_recv: Receiver<ContactForceEvent>,
+    // pub event_handler: ChannelEventCollector,
```
Comment (rasmusgo) [inline hotham/src/contexts/physics_context.rs:24]: Why is this removed?

**Comment by rasmusgo** [comment]
There is dark arc on the flat area close to the spheres on the side towards the light in the normal tangents test. It looks weird and wasn't there before. Can it have anything to do with ambient occlusion? Spotlights should not be affected by ambient occlusion...

File: exporters/export.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+# I will face Python.
+# I will permit it to pass over me and through me.
+# And when it has gone past, I will turn the inner eye to see its path.
+# Where the Python has gone there will be nothing.
+# Only Rust will remain
+
+class SaveOnWrite:
+    def export(dummy):
+        print("Hello?")
+        bpy.ops.export_scene.gltf(export_format="GLB", filepath="C:\\Users\\kanem\\Development\\hotham\\test_assets\\hot_cube.glb")
```
Comment (rasmusgo) [inline exporters/export.py:27]: This hardcoded path is not great. Both the path to the repo and the filename are problematic.

File: hotham-asset-client/src/client.rs [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+pub async fn watch(asset_names: Vec<String>, sender: Sender<AssetUpdatedMessage>) -> Result<()> {
+    run_client(asset_names, sender).await
+}
+
+async fn run_client(
+    mut asset_names: Vec<String>,
+    sender: Sender<AssetUpdatedMessage>,
+) -> Result<()> {
+    let server_addr = "192.168.1.124:5000".parse().unwrap();
```
Comment (rasmusgo) [inline hotham-asset-client/src/client.rs:19]: This hardcoded address is problematic. Can we use a parameter or an environment variable instead?

File: hotham/src/contexts/physics_context.rs [hunk 5]
```diff
@@ -1,4 +1,3 @@
-use crossbeam::channel::Receiver;
```
Comment (rasmusgo) [inline hotham/src/contexts/physics_context.rs:1]: This can be replaced with `rapier3d::crossbeam` if we don't want to depend on `crossbeam` directly.

**Comment by rasmusgo** [comment]
It doesn't look like `hotham/src/shaders/all_textures_and_ibl.frag` or `hotham/src/shaders/all_textures_no_ibl.frag` are ever used, are they?

**Comment by kanerogers** [comment]
> The gui render doesn't look good. The color of the ferris crab is broken, see `test_assets/render_draw_gui.jpg` image

Ah! Good pickup. I forgot to check these in on my desktop machine - have fixed these now.

File: Cargo.toml (hunk 1 above)
Comment (kanerogers) [reply Cargo.toml:8]: Hrm, let me fix that up.

File: hotham/src/contexts/physics_context.rs (hunk 2 above)
Comment (kanerogers) [reply hotham/src/contexts/physics_context.rs:24]: Good question!

- We never actually use Rapier events. We probably should, at some point in the future, because they're more efficient
- By removing them we can avoid an unnecessary crossbeam dependency

**Comment by kanerogers** [comment]
> 



> It doesn't look like `hotham/src/shaders/all_textures_and_ibl.frag` or `hotham/src/shaders/all_textures_no_ibl.frag` are ever used, are they?

Nope! They're redundant - I'll kill 'em with fire.

File: exporters/export.py (hunk 3 above)
Comment (kanerogers) [reply exporters/export.py:27]: Yeah, absolutely. This script was really just an experiment - I'll kill it off for now.

File: hotham-asset-client/src/client.rs (hunk 4 above)
Comment (kanerogers) [reply hotham-asset-client/src/client.rs:19]: Yeah I went back and forth on a couple of approaches for this. We're missing a good general solution for configuration settings like these!

For now I've stuck to `option_env` which should check on the host machine for an environment variable. I made this optional as it would be extremely annoying to make this mandatory, and the error message will bubble back up to the engine if anybody is looking. I also made it a *host* variable as setting environment variables on Android is possible, but annoying (again, it would be good to have a good general solution for this).

**Comment by kanerogers** [comment]
> There is dark arc on the flat area close to the spheres on the side towards the light in the normal tangents test. It looks weird and wasn't there before. Can it have anything to do with ambient occlusion? Spotlights should not be affected by ambient occlusion...
> 
> 

- re: ambient occlusion, I had to check the spec, but you are right! That's a bug - I've created an issue for it #416 
- But that's not an issue, as there's no ambient occlusion texture on that model
- Instead I think the issue is likely related to the (significantly faster) approximations we're using for the BRDFs, particularly the diffuse term, which I've mostly lifted them from Filament. For a real world scene, I haven't noticed any discernable difference, but given that we were originally using *the most accurate* BRDFs we could (lifted from a model viewer, rather than a real time engine), we will absolutely see some drop in accuracy in exchange for performance.
- I think this will always be a moving target between performance / quality. In the future it'd be good to give developers some more knobs they can tweak, but for now I believe this is an acceptable trade-off!

**Comment by kanerogers** [comment]
Strangely, clippy in CI seems to be applying more lints than on my local machine. This is *very frustrating*..

**Comment by kanerogers** [comment]
done

**Comment by rasmusgo** [comment]
The custom rendering example is broken 😬

File: Cargo.toml [hunk 6]
```diff
@@ -1,13 +1,16 @@
...
-    "examples/crab-saber",
-    "benchmarks/stress-test",
+  "hotham",
+  "hotham-simulator",
+  "examples/custom-rendering",
+  "examples/shared",
+  "examples/complex-scene",
+  "examples/simple-scene",
+  "examples/complex-scene",
```
Comment (rasmusgo) [inline Cargo.toml:9]: `complex-scene` is listed twice

**Review by rasmusgo — APPROVED** [review]
